### DISCUSSION TITLE
How to parse PKT_ TYPE_ DATA

### DISCUSSION THREAD

**Discussion by gemone** [FreeRDP/FreeRDP#8036] [General]
Hi all,

Sorry to disturb you. I am studying the source code recently. 

There are some problems in the gateway.

In [MS-TSGU/HTTP_DATA_PACKET Structure](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tsgu/5993d7bb-c0dd-4800-878a-e52faecffd25), does the `data (variable)` the event or bitmap change transmitted by RDP?

I try to find relevant parsing methods in the source code. But I found it:

```c
// libfreerdp/core/gateway/rdg.c 
// static int rdg_read_data_packet(rdpRdg* rdg, BYTE* buffer, int size); (line 2264)
		if (header.type != PKT_TYPE_DATA)
		{
			status = rdg_process_control_packet(rdg, header.type, header.packetLength);

			if (!status)
				return -1;

			return 0;
		}

		readCount = 0;

		while (readCount < 2)
		{
			status =
			    rdg_socket_read(rdg->tlsOut->bio, (BYTE*)(&rdg->packetRemainingCount) + readCount,
			                    2 - (int)readCount, &rdg->transferEncoding);

			if (status < 0)
			{
				if (!BIO_should_retry(rdg->tlsOut->bio))
					return -1;

				BIO_wait_read(rdg->tlsOut->bio, 50);
				continue;
			}

			readCount += (size_t)status;
		}
```

Is there a way to parse the data here?

And I found that `rdg_ bio_ read` reference modification method.

![image](https://user-images.githubusercontent.com/53191371/177241250-589454a8-ef88-4d00-b5f5-2674e27b8c40.png)

```c
// line 2519
static BIO_METHOD* BIO_s_rdg(void)
{
	static BIO_METHOD* bio_methods = NULL;

	if (bio_methods == NULL)
	{
		if (!(bio_methods = BIO_meth_new(BIO_TYPE_TSG, "RDGateway")))
			return NULL;

		BIO_meth_set_write(bio_methods, rdg_bio_write);
		BIO_meth_set_read(bio_methods, rdg_bio_read);
		BIO_meth_set_puts(bio_methods, rdg_bio_puts);
		BIO_meth_set_gets(bio_methods, rdg_bio_gets);
		BIO_meth_set_ctrl(bio_methods, rdg_bio_ctrl);
		BIO_meth_set_create(bio_methods, rdg_bio_new);
		BIO_meth_set_destroy(bio_methods, rdg_bio_free);
	}

	return bio_methods;
}
```

But `BIO_s_rdg` just return `BIO_METHOD*`.

```c
// line 2615 rdpRdg* rdg_new(rdpContext* context);
		rdg->frontBio = BIO_new(BIO_s_rdg());
```

And write method: 

```c
// line 2093
static int rdg_write_data_packet(rdpRdg* rdg, const BYTE* buf, int isize)
{
	if (rdg->transferEncoding.isWebsocketTransport)
	{
		if (rdg->transferEncoding.context.websocket.closeSent == TRUE)
			return -1;
		return rdg_write_websocket_data_packet(rdg, buf, isize);
	}
	else
		return rdg_write_chunked_data_packet(rdg, buf, isize);
}
```

Where is the `BYTE* buf` come from?

Only relevant definitions are found, but no references are found. Is there any relevant explanation.

Has anyone done relevant analysis? Can it be parsed into a frame?

> I try to at the gateway get `PKT_ TYPE_ DATA` intercepts frames to save user's RDP session. Is this feasible?

Thanks!

**Comment by akallabeth** [comment]
> In [MS-TSGU/HTTP_DATA_PACKET Structure](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tsgu/5993d7bb-c0dd-4800-878a-e52faecffd25), does the data (variable) the event or bitmap change transmitted by RDP?

what do you mean with that? (sorry, don´t understand what you try to tell)

> Is there a way to parse the data here?

You can add parsing code but to what end?

> I try to at the gateway get PKT_ TYPE_ DATA intercepts frames to save user's RDP session. Is this feasible?

Well, yes, but that traffic will be a SSL/TLS encrypted data stream.
For access to the actual unencrypted RDP session look at `stream_dump_register_handlers`

**Reply by gemone** [reply]
Thanks.

> what do you mean with that? (sorry, don´t understand what you try to tell)

I want to ask what is the specific data transmitted by `PKT_TYPE_DATA`. Are user events or changes in bitmap data?

> You can add parsing code but to what end?

I try to capture this part of data in the self implemented gateway to record the user's behavior.

The RDG like: [bolkedebruin/rdpgw](https://github.com/bolkedebruin/rdpgw).

[code here: ](https://github.com/bolkedebruin/rdpgw/blob/281cf1283ca140f936b9fbd5f2bf88cbb4f207f3/cmd/rdpgw/protocol/server.go#L145)

```go
		case PKT_TYPE_DATA:
			if s.State < SERVER_STATE_CHANNEL_CREATE {
				log.Printf("Data received while in wrong state %d != %d", s.State, SERVER_STATE_CHANNEL_CREATE)
				return errors.New("wrong state")
			}
			s.State = SERVER_STATE_OPENED
                       
                        // Define the recording method
                        capturePKT(pkt)
			receive(pkt, s.Remote)
```

For `capture session`, will this implementation be easier than proxy? The gateway follows the `MS-TGSU` specification and does not need to consider the forwarding of each traffic.

> At present, the proxy cannot connect to the target server of NLA authentication.

**Reply by akallabeth** [reply]
1. All `RDP` data that is tunneled over the gateway is transmitted there. This is data encapsulated in a `SSL/TLS` layer. (everything else would be a gaping security hole)
2. No, you will not be able to capture something useful there. (again, `SSL/TLS` encrypted data)
3. What problems with `NLA` do you have? (note: in current `2.0` releases `proxy` is `experimental`, so you better switch to `master` where there have been huge improvements regarding that, still not stable but if you are already on the edge ;))

**Reply by gemone** [reply]
> 3\. What problems with `NLA` do you have? (note: in current `2.0` releases `proxy` is `experimental`, so you better switch to `master` where there have been huge improvements regarding that, still not stable but if you are already on the edge ;))

Is `proxy server` depends the ` client` ?

`freerdp client` -> `freerdp proxy` -> `target server`

but

`mstsc` -x> `freerdp proxy`

**Reply by akallabeth** [reply]
should work with `mstsc`, I´ve used it that way. so maybe some setup issue on your end?

**Reply by gemone** [reply]
Thanks.

Maybe my configuration is wrong. I'll try to modify it here.

If `NLA` is enabled, `xfreerdp` cannot establish a connection, but `mstsc` can. If `NLA` is turned off, `xfreerdp` can connect, but mstsc cannot.

```ini
[Security]
ServerNlaSecurity=true
```

> `xfreerdp 2.7` cannot connect. But `master` can.

```ini
[Server]
Host=0.0.0.0
Port=3389

[Target]
Host=<target ip>
Port=3389
FixedTarget=true

[Channels]
GFX=true
DisplayControl=true
Clipboard=true
AudioInput=true
AudioOutput=true
DeviceRedirection=true
VideoRedirection=true
CameraRedirection=true
RemoteApp=false
PassthroughIsBlacklist=true
Passthrough=
Intercept=

[Input]
Keyboard=true
Mouse=true
Multitouch=true

[Security]
ServerTlsSecurity=true
ServerNlaSecurity=true
ServerRdpSecurity=true
ClientTlsSecurity=true
ClientNlaSecurity=true
ClientRdpSecurity=true
ClientAllowFallbackToTls=true

[Plugins]
Modules=
Required=

[Clipboard]
TextOnly=false
MaxTextLength=0

[GFXSettings]
DecodeGFX=false

[Certificates]
CertificateFile=./certs/rdpgw.crt
PrivateKeyFile=./certs/rdpgw.key
RdpKeyFile=./certs/rdpgw.key
```
